libs/types/src/tokens.rs

```diff
@@ -322,17 +375,116 @@ mod tests {
 			))
 			.is_err()
 		);
-		assert!(
-			TryInto::<GeneralCurrencyIndex<u128, ZeroPrefix>>::try_into(CurrencyId::KSM).is_err()
-		);
-		assert!(
-			TryInto::<GeneralCurrencyIndex<u128, ZeroPrefix>>::try_into(CurrencyId::AUSD).is_err()
-		);
 		assert!(
 			TryInto::<GeneralCurrencyIndex<u128, ZeroPrefix>>::try_into(CurrencyId::Staking(
 				StakingCurrency::BlockRewards
 			))
 			.is_err()
 		);
 	}
+
+	#[cfg(test)]
+	mod tests {
+		use cfg_primitives::TrancheId;
+		use codec::Encode;
+		use hex::FromHex;
+
+		use super::StakingCurrency;
+		use crate::{tokens as after, tokens::before};
+
+		#[test]
+		fn currency_id_refactor_encode_equality() {
+			// Native
+			assert_eq!(
+				before::CurrencyId::Native.encode(),
+				after::CurrencyId::Native.encode()
+			);
+			assert_eq!(after::CurrencyId::Native.encode(), vec![0]);
+
+			// Tranche
+			assert_eq!(
+				before::CurrencyId::Tranche(33, default_tranche_id()).encode(),
+				after::CurrencyId::Tranche(33, default_tranche_id()).encode()
+			);
+			assert_eq!(
+				after::CurrencyId::Tranche(33, default_tranche_id()).encode(),
+				vec![
+					1, 33, 0, 0, 0, 0, 0, 0, 0, 129, 26, 205, 91, 63, 23, 192, 104, 65, 199, 228,
+					30, 158, 4, 203, 27
+				]
+			);
+
+			// KSM - deprecated
+			assert_eq!(before::CurrencyId::KSM.encode(), vec![2]);
+
+			// AUSD - deprecated
+			assert_eq!(before::CurrencyId::AUSD.encode(), vec![3]);
+
+			// ForeignAsset
+			assert_eq!(
+				before::CurrencyId::ForeignAsset(91).encode(),
+				after::CurrencyId::ForeignAsset(91).encode()
+			);
+			assert_eq!(
+				after::CurrencyId::ForeignAsset(91).encode(),
+				vec![4, 91, 0, 0, 0]
+			);
+
+			// Staking
+			assert_eq!(
+				before::CurrencyId::Staking(StakingCurrency::BlockRewards).encode(),
+				after::CurrencyId::Staking(StakingCurrency::BlockRewards).encode()
+			);
+			assert_eq!(
+				after::CurrencyId::Staking(StakingCurrency::BlockRewards).encode(),
+				vec![5, 0]
+			);
+		}
+
+		fn default_tranche_id() -> TrancheId {
+			<[u8; 16]>::from_hex("811acd5b3f17c06841c7e41e9e04cb1b")
+				.expect("Should be valid tranche id")
+		}
+	}
+
+	/// Sanity check for every CurrencyId variant's encoding value.
+	/// This will stop us from accidentally moving or dropping variants
+	/// around which could have silent but serious negative consequences.
+	#[test]
+	fn currency_id_encode_sanity() {
+		// Verify that every variant encodes to what we would expect it to.
+		// If this breaks, we must have changed the order of a variant, added
+		// a new variant in between existing variants, or deleted one.
+		vec![
+			Native,
+			Tranche(42, [42; 16]),
+			AUSD,
```

Comment:
AFAIK, we must add `KSM` here as well. Moreover, how do you envision the test to stay in its current form if we remove `AUSD` from the `CurrencyId` enum? It appears to me that this test is not foolproof.

Reply:
No, KSM can already and is already dropped since we don't have to migrate anything for KSM.

>  It appears to me that this test is not foolproof.

It's not, indeed. This test has value if we don't go in messing around with the CurrencyId variants, which is exactly what we did in this PR (removing AUSD and then adding it back to make the migration possible).

Once we remove AUSD in a follow up PR (after the migration is applied), we drop the `CurrencyId::AUSD` variant and drop `AUSD` from this list.